packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/internal/SubscriptionImpl.kt

```diff
@@ -0,0 +1,80 @@
+package io.realm.mongodb.internal
+
+import io.realm.BaseRealm
+import io.realm.RealmInstant
+import io.realm.RealmObject
+import io.realm.TypedRealm
+import io.realm.internal.RealmInstantImpl
+import io.realm.internal.interop.RealmBaseSubscriptionSetPointer
+import io.realm.internal.interop.RealmInterop
+import io.realm.internal.interop.RealmSubscriptionPointer
+import io.realm.mongodb.sync.Subscription
+import io.realm.query.RealmQuery
+import kotlin.reflect.KClass
+
+internal class SubscriptionImpl(
+    private val realm: BaseRealm,
+    private val parentNativePointer: RealmBaseSubscriptionSetPointer,
+    internal val nativePointer: RealmSubscriptionPointer
+) : Subscription {
+    override val createdAt: RealmInstant = RealmInstantImpl(RealmInterop.realm_sync_subscription_created_at(nativePointer))
+    override val updatedAt: RealmInstant = RealmInstantImpl(RealmInterop.realm_sync_subscription_updated_at(nativePointer))
+    override val name: String? = RealmInterop.realm_sync_subscription_name(nativePointer)
+    override val objectType: String = RealmInterop.realm_sync_subscription_object_class_name(nativePointer)
+    override val queryDescription: String = RealmInterop.realm_sync_subscription_query_string(nativePointer)
+
+    @Suppress("invisible_member")
+    override fun <T : RealmObject> asQuery(type: KClass<T>): RealmQuery<T> {
+        // TODO Check for invalid combinations of Realm and type once we properly support
+        // DynamicRealm
+        return when (realm) {
+            is TypedRealm -> {
+                val companion = io.realm.internal.platform.realmObjectCompanionOrThrow(type)
+                val userTypeName = companion.`io_realm_kotlin_className`
+                if (userTypeName != objectType) {
+                    throw IllegalArgumentException(
+                        "Wrong query type. This subscription is for " +
+                            "objects of type: $objectType, but $userTypeName was provided as input."
+                    )
+                }
+                realm.query(type, queryDescription)
+            }
+            // is DynamicRealm -> {
+            //     if (type != DynamicRealmObject::class) {
+            //         throw IllegalArgumentException(
+            //             "This subscription was fetched from a " +
+            //                 "DynamicRealm, so the type argument must be `DynamicRealmObject`."
+            //         )
+            //     }
+            //     realm.query(className = objectType, query = queryDescription) as RealmQuery<T>
+            // }
+            else -> {
+                TODO("Unsupported Realm type: ${realm::class}")
+            }
+        }
+    }
+
```

Comment:
Hmm, I think `IllegalArgumentException` make sense here, since all of this code is derived from the `type` argument the user comes with, so if something goes wrong here, it is because the type was illegal.